kernel/src/main/java/org/kframework/kompile/DefinitionAndCache.java

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2019 K Team. All Rights Reserved.
+package org.kframework.kompile;
+
+import org.kframework.parser.concrete2kore.ParseCache;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Represents compiled definition, parse cache and extended definitions derived from this base definition at runtime.
+ *
+ * @author Denis Bogdanas
+ * Created on 05-Nov-19.
+ */
+public class DefinitionAndCache implements Serializable {
+    public final @Nullable CompiledDefinition compiledDefinition;
+    public final Map<String, ParseCache> parseCaches;
+
+    public DefinitionAndCache(CompiledDefinition compiledDefinition,
+                              Map<String, ParseCache> parseCaches) {
+        this.compiledDefinition = compiledDefinition;
+        this.parseCaches = Collections.synchronizedMap(parseCaches);
```

Comment:
I added synchronization just for the outer map, here.

Reply:
Let's make it a ConcurrentHashMap instead of a synchronizedMap. I think you're right that with the increased parallelism I added to the codebase recently, in theory this can be accessed from multiple threads, but we definitely need lockless reading from the map otherwise there will be a huge amount of contention between threads. It should be fine performance-wise if we make this change, because writing to the outer map happens at most once per module.